File: 8-TapFlyAndActiveTrack/P4Missions/RootViewController.m
//
//  RootViewController.m
//  P4Missions
//
//  Created by DJI on 15/3/16.
//  Copyright © 2016 DJI. All rights reserved.
//

#import "RootViewController.h"
#import "DemoUtility.h"

#define ENTER_DEBUG_MODE 0

@interface RootViewController ()<DJISDKManagerDelegate>
@property (weak, nonatomic) IBOutlet UIButton *tapFlyMissionButton;
@property (weak, nonatomic) IBOutlet UIButton *activeTrackMissionButton;

@end

@implementation RootViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    self.title = @"TapFly & ActiveTrack Missions Demo";
    [self registerApp];
    
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark Custom Methods

- (void)registerApp
{
    //Please enter the App Key in the info.plist file to register the app.
    [DJISDKManager registerAppWithDelegate:self];
}

- (void)showAlertViewWithTitle:(NSString *)title withMessage:(NSString *)message
{
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:okAction];
    [self presentViewController:alert animated:YES completion:nil];
    
}

#pragma mark DJISDKManagerDelegate Method

- (void)productConnected:(DJIBaseProduct *)product
{
    if (product) {
        [self.tapFlyMissionButton setEnabled:YES];
        [self.activeTrackMissionButton setEnabled:YES];
    }else
    {
        [self.tapFlyMissionButton setEnabled:NO];
        [self.activeTrackMissionButton setEnabled:NO];
    }
    
    //If this demo is used in China, it's required to login to your DJI account to activate the application. Also you need to use DJI Go app to bind the aircraft to your DJI account. For more details, please check this demo's tutorial.
    [[DJISDKManager userAccountManager] logIntoDJIUserAccountWithAuthorizationRequired:NO withCompletion:^(DJIUserAccountState state, NSError * _Nullable error) {
        if (error) {
            ShowResult(@"Login failed: %@", error.description);
        }
    }];
}

- (void)productDisconnected
{
    [self.tapFlyMissionButton setEnabled:NO];
    [self.activeTrackMissionButton setEnabled:NO];
}

- (void)appRegisteredWithError:(NSError *)error
{
    if (error) {
        NSString* message = @"Register App Failed! Please enter your App Key and check the network.";
        [self.tapFlyMissionButton setEnabled:NO];
        [self.activeTrackMissionButton setEnabled:NO];
        [self showAlertViewWithTitle:@"Register App" withMessage:message];

    }else
    {
        NSLog(@"registerAppSuccess");
#if ENTER_DEBUG_MODE
        [DJISDKManager enableBridgeModeWithBridgeAppIP:@"10.61.12.100"];
#else
        [DJISDKManager startConnectionToProduct];
#endif
    
    }
}

@end

File: 10-GEOSystem/DJIGeoSample/RootViewController.swift
//
//  RootViewController.swift
//  DJIGeoSample
//
//  Created by Samuel Scherer on 5/3/21.
//  Copyright © 2021 RIIS. All rights reserved.
//

import Foundation

import Foundation
import UIKit
import DJISDK

let kUseBridge = false
let kBridgeIP = "192.168.128.169"

class RootViewController : UIViewController, DJISDKManagerDelegate {
    var product : DJIBaseProduct?
    @IBOutlet weak var connectStatusLabel: UILabel!
    @IBOutlet weak var modelNameLabel: UILabel!
    @IBOutlet weak var connectButton: UIButton!
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        //Please enter the App Key in info.plist file to register the app.
        DJISDKManager.registerApp(with: self)
        if let product = self.product {
            self.updateStatusFor(product)
        }
    }

    override func viewDidLoad() {
        self.initUI()
        super.viewDidLoad()
    }

    func initUI() {
        self.title = "DJI GEO Demo"
        self.modelNameLabel.isHidden = true
        //Disable the connect button by default
        self.connectButton.isEnabled = false
    }
    
    func updateStatusFor(_ product:DJIBaseProduct?) {
        if let product = product {
            self.connectStatusLabel.text = "Status: Product Connected"
            self.modelNameLabel.text = "Model: \(product.model ?? "Unknown")"
            self.modelNameLabel.isHidden = false
        } else {
            self.connectStatusLabel.text = "Status: Product Not Connected"
            self.modelNameLabel.text = "Model: Unknown"
        }
    }

//MARK: - DJISDKManager Delegate Methods
    func appRegisteredWithError(_ error: Error?) {
        if let error = error {
            showAlertWith(result: "Registration Error: \(error)")
            self.connectButton.isEnabled = false
            return
        }
        if kUseBridge {
            DJISDKManager.enableBridgeMode(withBridgeAppIP: kBridgeIP)
        } else {
            DJISDKManager.startConnectionToProduct()
        }
    }

//MARK: DJIBaseProductDelegate Method
    func productConnected(_ product: DJIBaseProduct?) {
        if let product = product {
            self.product = product
            self.connectButton.isEnabled = true
        } else {
            self.productDisconnected()
        }
        self.updateStatusFor(product)
        
        //If this demo is used in China, it's required to login to your DJI account to activate the application. Also you need to use DJI Go app to bind the aircraft to your DJI account. For more details, please check this demo's tutorial.
        DJISDKManager.userAccountManager().logIntoDJIUserAccount(withAuthorizationRequired: false) { (state:DJIUserAccountState, error:Error?) in
            if let error = error {
                print("Login failed: \(error.localizedDescription)")
            }
        }
    }
    
    func productDisconnected() {
        let message = "Connection lost. Back to root."
        let cancelAction = UIAlertAction(title: "OK", style: UIAlertAction.Style.cancel, handler: nil)
        let backAction = UIAlertAction(title: "Back", style: UIAlertAction.Style.default) { (action:UIAlertAction) in
            if let topVC = self.navigationController?.topViewController {
                if !(topVC is RootViewController) {
                    self.navigationController?.popToRootViewController(animated: true)
                }
            }
        }
        showAlertWith(title: nil, message: message, cancelAction: cancelAction, defaultAction: backAction, presentingViewController: self)
        self.connectButton.isEnabled = false
        self.product = nil
        
        self.updateStatusFor(nil)
    }
    
    func didUpdateDatabaseDownloadProgress(_ progress: Progress) {}
}
